### ISSUE TITLE
coming should always write `coming_result.json` at the end of an analysis

### ISSUE THREAD

**Issue by monperrus** [SpoonLabs/coming#124]
the Coming core should collect all JSON results, process them and write them to disk in `coming_result.json`.

For this, it'd be nice to have FinalResult be serializable to JSON

**Comment by martinezmatias** [comment]
Hi Martin
Right now we have an extension point with one implementation of it that exports the results to JSON.
The interface that represents the extension point is: https://github.com/SpoonLabs/coming/blob/HEAD/src/main/java/fr/inria/coming/core/entities/interfaces/IOutput.java

and one of the implementation is : https://github.com/SpoonLabs/coming/blob/ecfcf05577afdd0ab7a1eeb19d0b471e6ec6f750/src/main/java/fr/inria/coming/core/entities/output/JSonPatternInstanceOutput.java

(Via command line we can pass severals output)

I agree that this part must be improved, e.g., I would add to the entities `FinalResults` and `RevisionResult` an abstract method `toJSON`. 
Now, the outputs are completely uncoupled from the analyzers and from the results. Maybe we would add an explicit link e.g., "always that I use an Analyzer X, then the output processor will be Y " .
WDYT?


WDYT?

**Comment by monperrus** [comment]
> I would add to the entities `FinalResults` and `RevisionResult` an abstract method `toJSON`.

Yes that would be great. The goal is that every run create by default a JSON file and the last line in the log would be "created 'coming_result.json'": this is the added value of Coming (instead of the log lines, see #115).

(The inspiration for this comes from the great https://github.com/mast-group/mineSStuBs/)

**Comment by martinezmatias** [comment]
> The goal is that every run create by default a JSON file and the last line in the log would be "created 'coming_result.json'"

Yes, actually we have something similar to that [(example)](https://github.com/SpoonLabs/coming/blob/038bdc08866ea90d89f55a8f0d21bf4b37d907bf/src/main/java/fr/inria/coming/core/entities/output/JSonPatternInstanceOutput.java#L61). However, as shown in that example, it's each output processor who stores the JSON (and eventually logs the path to the file). I would say that the Coming core must do that instead (i.e., the output processor should only return the json and the core processes them)

**Comment by monperrus** [comment]
Fully agree. Updated the title and desc of the PR.

@niloofartrg what about having a look at this in parallel of your experiments?

**Comment by monperrus** [comment]
after discussion with @niloofartrg 

it's the responsibility of each mode to save the JSON file at the right place.

#130 makes sure that the features mode do create one JSON file per commit
